### PR TITLE
Path substitution

### LINKED ISSUE #43: Add YAML path resolving to substitution
At the Royal Danish Library, multiple YAML files are used, each shadowing each other. A mechanism for substituting a value with a a value taken from another path in the hierarchy would make it easier to split definition of sensible information away from the YAML files where they are used.

Suggestion:

File 1:
```
servers:
  database:
    url: 'http://example.org/db'
```

File 2 (shadowing File 1):
```
db:
  threads: 10
  connection: ${path:servers.database.url}
```

### PR DESCRIPTION
**Pull request by tokee**
This pull request makes it possible to perform path substitution (simple values only - for now). Besides the example given in #43  it is possible to make a conditional substitution such as
```
databases:
  - production:
    -  url: http://example.com/prod"
  - test:
    -  url: http://example.com/test"
    - default: true

newservice:
  threads: 16
  databaseurl: ${path:databases[default=true].url}
```

This closes #43

### CODE REVIEW

File: src/test/java/dk/kb/util/yaml/YAMLTest.java [hunk 1]
```diff
@@ -528,4 +528,41 @@ private void assertExtrapolated(String path, String expected, String message) th
...
+        YAML yaml = YAML.resolveLayeredConfigs("nested_maps.yml");
+        assertEquals("bar", yaml.get("conditionalmap.[default!=true].foo"));
+    }
+
+    @Test
+    public void testConditionalIndexFlat() throws IOException {
+        YAML yaml = YAML.resolveLayeredConfigs("nested_maps.yml");
+        assertEquals("boom", yaml.get("conditionalflat.[default=true].foo"));
+    }
+
```
Comment (VictorHarbo) [inline src/test/java/dk/kb/util/yaml/YAMLTest.java:559]: Same as above, but for the conditionalflatmap

**Review by VictorHarbo — CHANGES_REQUESTED** [review]
Test question

File: src/test/java/dk/kb/util/yaml/YAMLTest.java [hunk 2]
```diff
@@ -528,4 +528,41 @@ private void assertExtrapolated(String path, String expected, String message) th
...
+    public void testConditionalIndex() throws IOException {
+        YAML yaml = YAML.resolveLayeredConfigs("nested_maps.yml");
+        assertEquals("boom", yaml.get("conditionalmap.[default=true].foo"));
+    }
+
+    @Test
+    public void testReverseConditionalIndex() throws IOException {
+        YAML yaml = YAML.resolveLayeredConfigs("nested_maps.yml");
+        assertEquals("bar", yaml.get("conditionalmap.[default!=true].foo"));
+    }
```
[the remark this answers is not in the document]
Comment (tokee) [reply src/test/java/dk/kb/util/yaml/YAMLTest.java]: The YAML structure used for testing was a poor design. I have extended the unit test to 4 different ways of presenting multiple maps, including one using the clean design where `conditionalpropermap.bucket2.foo` works (a unit test has been added for this exact case).

File: src/test/java/dk/kb/util/yaml/YAMLTest.java (hunk 1 above)
Comment (tokee) [reply src/test/java/dk/kb/util/yaml/YAMLTest.java]: Same answer as above :-)

**Comment by tokee** [comment]
Pair-review with @jorntx completed for the fix for the issue raised by @VictorHarbo . Jørn told me to merge.

**Pull request merged**